### PR TITLE
More intuition on tree construction

### PR DESCRIPTION
**Pull request by felixlinker**
With this PR, I try to give a more high-level intuition of what happens in KT, but there is a discrepancy between what I write here (and the Merkle^2 paper) and what you write. I will open an issue separately for that.

### CODE REVIEW

File: draft-keytrans-mcmillion-protocol.md [hunk 1]
```diff
@@ -78,10 +78,23 @@ properties are maintained.
 
 # Tree Construction
 
-KT relies on two combined hash tree structures: log trees and prefix trees. This
-section describes the operation of both at a high level and the way that they're
-combined. More precise algorithms for computing the intermediate and root values
-of the trees are given in {{cryptographic-computations}}.
+KT allows clients of a service to query the keys of other clients of the same
+service. To do so, KT maintains two types of maps: (i) one associates *version
```
Comment (Bren2010) [inline draft-keytrans-mcmillion-protocol.md:82]: Maintains two things:
- A log of changes to lookup keys
- A set of key-version pairs

File: draft-keytrans-mcmillion-protocol.md [hunk 2]
```diff
@@ -197,8 +214,8 @@ added.
 
 In the combined tree structure, which is based on {{Merkle2}}, a log tree
 maintains a record of each time any key's value is updated, while a prefix tree
-maintains the set of key-version pairs. Importantly, the root value of the
-prefix tree after adding the new key-version pair is stored in a leaf of the log
+maintains the set of pseudonym-version pairs. Importantly, the root hash value of the
+prefix tree after adding a new pseudonym-version pair is stored in a leaf of the log
```
Comment (Bren2010) [inline draft-keytrans-mcmillion-protocol.md:218]:
```suggestion
maintains the set of index-version pairs. Importantly, the root hash value of the
prefix tree after adding a new index-version pair is stored in a leaf of the log
```
Need to use index consistently, not pseudonym

File: draft-keytrans-mcmillion-protocol.md [hunk 3]
```diff
@@ -130,59 +148,58 @@ log.
...
+all values that share its prefix followed by an additional 0 bit, while the
+subtree headed by a parent's right child contains all values that share its
+prefix followed by an additional 1 bit.
 
 A prefix tree can be searched by starting at the root node, and moving to the
-left child if the first bit of a value is 0, or the right child if the first bit
-is 1. This is then repeated for the second bit, third bit, and so on until the
-search either terminates at a leaf node (which may or may not be for the desired
-value), or a parent node that lacks the desired child.
+left child if the first bit of an index is 0, or the right child if the first
```
Comment (Bren2010) [inline draft-keytrans-mcmillion-protocol.md:175]:
```suggestion
left child if the first bit of a value is 0, or the right child if the first
```
"Prefix tree is used for storing a set of values" so sticking with value

File: draft-keytrans-mcmillion-protocol.md [hunk 4]
```diff
@@ -78,16 +78,30 @@ properties are maintained.
...
-
-Both types of trees consist of *nodes* which have a byte string as their
-*value*. A node is either a *leaf* if it has no children, or a *parent* if it
-has either a *left child* or a *right child*. A node is the *root* of a tree if
-it has no parents, and an *intermediate* if it has both children and parents.
-Nodes are *siblings* if they share the same parent.
+KT allows clients of a service to query the keys of other clients of the same
+service. To do so, KT maintains two maps: (i) one associates key-update counters
+with key updates, (ii) the other associates anonymized pairs of pseudonyms and
+version numbers with key-update counters. When clients query a KT service, they
```
[the remark this answers is not in the document]
Comment (felixlinker) [reply draft-keytrans-mcmillion-protocol.md:84]: I think it's important to point out how the leafs of each tree are addressed. For (i), it's by a global version counter (actually, we should find a name for this; is there one?). For (ii), it's the output of the VRF for pseudonym + local version counter. I find this very important! A set cannot be indexed, yet we will index the trees.

File: draft-keytrans-mcmillion-protocol.md [hunk 5]
```diff
@@ -101,7 +115,10 @@ of all the nodes in its direct path, excluding the root.
 ## Log Tree
 
 Log trees are used for storing information in the chronological order that it
-was added and are constructed as *left-balanced* binary trees.
+was added and are constructed as *left-balanced* binary trees. The log tree that
+we utilize here will have incrementing counters as its indices and pairs of
+key-updates and prefix-tree hashes as its values. The log tree's leaves are
+ordered by their respective index (counter value) ascending from left to right.
```
[the remark this answers is not in the document]
Comment (felixlinker) [reply draft-keytrans-mcmillion-protocol.md:120]: I think it's important to provide intuitions as early as possible. Here one could say that an intuition was already provided in the introduction of the combined tree structure earlier, but readers who skip between sections might miss this.

File: draft-keytrans-mcmillion-protocol.md [hunk 6]
```diff
@@ -114,14 +131,15 @@ child.
...
+added to the right-most edge of the tree along with a single parent node, to
+construct the left-balanced binary tree with `n+1` leaves.
 
 <!-- TODO diagram of adding a new leaf to a tree (specifically 5 leaves to 6 leaves to show level-skipping) -->
 
-While leaves contain arbitrary data, the value of a parent node is always the
-hash of the combined values of its left and right children.
+The hash value of a leaf node is a hash of its value, and the hash value of a
+parent node is the hash of the combined hash values of its left and right
+children.
```
[the remark this answers is not in the document]
Comment (felixlinker) [reply draft-keytrans-mcmillion-protocol.md:141]: Some complications arise from refactoring "value" into "hash value" and this is one of them. It was clearer before, but if we want to distinguish "hash value" from "value" (which I think we should), I'd rather be explicit here.

File: draft-keytrans-mcmillion-protocol.md [hunk 7]
```diff
@@ -130,59 +148,58 @@ log.
...
-that the result equals the root value of the log. Consistency proofs are a more
-general version of the same idea. With a consistency proof, the prover provides
-the minimum set of intermediate node values from the current tree that allows
-the verifier to compute both the old root value and the current root value. An
-algorithm for this is given in section 2.1.2 of {{!RFC6962}}.
+that the result equals the root hash value of the log. Consistency proofs are a
+more general version of the same idea. With a consistency proof, the prover
+provides the minimum set of intermediate node values from the current tree that
+allows the verifier to compute both the old root value and the current root
+value. An algorithm for this is given in section 2.1.2 of {{!RFC6962}}.
```
[the remark this answers is not in the document]
Comment (felixlinker) [reply draft-keytrans-mcmillion-protocol.md:154]: Do you have any preference regarding how we handle formatting? When would you like to enforce the 80 characters limit?

File: draft-keytrans-mcmillion-protocol.md [hunk 8]
```diff
@@ -130,59 +143,58 @@ log.
...
+prefix trees that we utilize here will have anonymized pairs of pseudonyms and
+version counters as its indices and the log tree's indices as values.
+
+The prefix tree is ordered by the shortest prefixes of its leaves. The prefix
+tree's root represents the empty prefix. All indices of the root's left subtree
+begin with a 0 bit, all indices of the root's right subtree start with a 1 bit.
+More generally speaking. The subtree headed by a parent's left child contains
+all values that share its prefix followed by an additional 0 bit, while the
+subtree headed by a parent's right child contains all values that share its
+prefix followed by an additional 1 bit.
```
Comment (Bren2010) [inline draft-keytrans-mcmillion-protocol.md:167]:
```suggestion
to efficiently produce proofs of membership and non-membership in the set.

Each leaf node in a prefix tree represents a specific value, while each parent
node represents some prefix which all values in the subtree headed by that node
have in common. The subtree headed by a parent's left child contains all values
that share its prefix followed by an additional 0 bit, while the subtree headed
by a parent's right child contains all values that share its prefix followed by
an additional 1 bit.

The root node, in particular, represents the empty string as a prefix. The
root's left child contains all values that begin with a 0 bit, while the right
child contains all values that begin with a 1 bit.
```

File: draft-keytrans-mcmillion-protocol.md [hunk 9]
```diff
@@ -130,59 +143,58 @@ log.
...
-leaf and the existing leaf would no longer reside in the same place. That is,
-until we reach the first bit that differs between the new value and the existing
-value.
+New index-value pairs are added to the tree by searching it according to the
+same process. If the search terminates at a parent without a left or right
+child, a new leaf is added as the parent's missing child. If the search
+terminates at a leaf for the wrong index, one or more intermediate nodes are
+added until the new leaf and the existing leaf would no longer reside in the
+same place. That is, until we reach the first bit that differs between the new
+index and the existing index.
```
Comment (Bren2010) [inline draft-keytrans-mcmillion-protocol.md:183]:
```suggestion
New values are added to the tree by searching it according to the same process.
If the search terminates at a parent without a left or right child, a new leaf
is simply added as the parent's missing child. If the search terminates at a
leaf for the wrong value, one or more intermediate nodes are added until the new
leaf and the existing leaf would no longer reside in the same place. That is,
until we reach the first bit that differs between the new value and the existing
value.
```

File: draft-keytrans-mcmillion-protocol.md [hunk 10]
```diff
@@ -130,59 +143,58 @@ log.
...
 
 A prefix tree can be searched by starting at the root node, and moving to the
-left child if the first bit of a value is 0, or the right child if the first bit
-is 1. This is then repeated for the second bit, third bit, and so on until the
-search either terminates at a leaf node (which may or may not be for the desired
-value), or a parent node that lacks the desired child.
+left child if the first bit of a value is 0, or the right child if the first
+bit is 1. This is then repeated for the second bit, third bit, and so on until
+the search either terminates at a leaf node (which may or may not be for the
+desired value), or a parent node that lacks the desired child.
```
Comment (Bren2010) [inline draft-keytrans-mcmillion-protocol.md:173]:
```suggestion
left child if the first bit of a value is 0, or the right child if the first bit
is 1. This is then repeated for the second bit, third bit, and so on until the
search either terminates at a leaf node (which may or may not be for the desired
value), or a parent node that lacks the desired child.
```

File: draft-keytrans-mcmillion-protocol.md [hunk 11]
```diff
@@ -130,59 +143,58 @@ log.
...
+index and the existing index.
 
 <!-- TODO diagram of adding new value -->
 
-The value of a leaf node is the encoded set member, while the value of a
-parent node is the hash of the combined values of its left and right children
-(or a stand-in value when one of the children doesn't exist).
+The hash value of a leaf node is the hash of its value, while the hash value of
+a parent node is the hash of the combined hash values of its left and right
+children (or a stand-in value when one of the children doesn't exist).
```
Comment (Bren2010) [inline draft-keytrans-mcmillion-protocol.md:189]:
```suggestion
The value of a leaf node is the encoded set member, while the value of a
parent node is the hash of the combined values of its left and right children
(or a stand-in value when one of the children doesn't exist).
```

**Review by Bren2010 — APPROVED** [review]

**Pull request merged**